sources/npmRegistryUtils.ts

```diff
@@ -32,30 +34,58 @@ export async function fetchAsJson(packageName: string, version?: string) {
   return httpUtils.fetchAsJson(`${npmRegistryUrl}/${packageName}${version ? `/${version}` : ``}`, {headers});
 }
 
-export function verifySignature({signatures, integrity, packageName, version}: {
+const getVerificationKeys = once(async (): Promise<Array<{ keyid: string, key: KeyObject }>> => {
+  let keys: Array<{ keyid: string, key: string }>;
+  if (process.env.COREPACK_INTEGRITY_KEYS) {
+    // We use the format of the `GET /-/npm/v1/keys` endpoint with `npm` instead
+    // of `keys` as the wrapping key.
+    const keysFromEnv = JSON.parse(process.env.COREPACK_INTEGRITY_KEYS) as { npm: Array<{ keyid: string, key: string }> };
+    keys = keysFromEnv.npm.map(k => ({keyid: k.keyid, key: k.key}));
+    debugUtils.log(`Using COREPACK_INTEGRITY_KEYS to verify signatures: ${keys.map(k => k.keyid).join(`, `)}`);
+  } else {
+    // This follows the implementation for npm.
+    // See https://github.com/npm/cli/blob/3a80a7b7d168c23b5e297cba7b47ba5b9875934d/lib/utils/verify-signatures.js#L174
+    // We only support sigstore for NPM. For other registries
+    // COREPACK_INTEGRITY_KEYS can be used.
+    const sigstoreTufClient = await sigstoreTuf.initTUF();
+    const keysRaw = await sigstoreTufClient.getTarget(`registry.npmjs.org/keys.json`);
+    // The format of the key file is undocumented, unfortunately. `rawBytes` is
+    // the PEM content, i.e. base64 encoded SPKI DER.
+    const keysFromSigstore = JSON.parse(keysRaw) as { keys: Array<{ keyId: string, publicKey: { rawBytes: string }}> };
+    keys = keysFromSigstore.keys.map(k => ({keyid: k.keyId, key: k.publicKey.rawBytes}));
+    debugUtils.log(`Using NPM keys from @sigstore/tuf to verify signatures: ${keys.map(k => k.keyid).join(`, `)}`);
+  }
+
+  return keys.map(k => ({
+    keyid: k.keyid,
+    key: createPublicKey(`-----BEGIN PUBLIC KEY-----\n${k.key}\n-----END PUBLIC KEY-----`,
+    )}));
+});
+
+export async function verifySignature({signatures, integrity, packageName, version}: {
   signatures: Array<{keyid: string, sig: string}>;
   integrity: string;
   packageName: string;
   version: string;
 }) {
-  const {npm: keys} = process.env.COREPACK_INTEGRITY_KEYS ?
-    JSON.parse(process.env.COREPACK_INTEGRITY_KEYS) as typeof defaultConfig.keys :
-    defaultConfig.keys;
-
+  const keys = await getVerificationKeys();
   const key = keys.find(({keyid}) => signatures.some(s => s.keyid === keyid));
-  const signature = signatures.find(({keyid}) => keyid === key?.keyid);
+  if (key == null)
+    throw new Error(`Cannot find key to verify signature. signature keys: ${signatures.map(s => s.keyid)}, verification keys: ${keys.map(k => k.keyid)}`);
 
-  if (key == null || signature == null) throw new Error(`Cannot find matching keyid: ${JSON.stringify({signatures, keys})}`);
+  const signature = signatures.find(({keyid}) => keyid === key.keyid);
+  assert(signature); // If `key` is defined there is a matching signature
```

Comment:
this comment could be clearer

Reply:
I would tend to remove the comment.

The comment is really about the interaction between these 3 lines:

```
// find a key that appears to produce a signature
const key = keys.find(({keyid}) => signatures.some(s => s.keyid === keyid));
// find the first signature produced by the key
const signature = signatures.find(({keyid}) => keyid === key.keyid);
// ensure there exists a signature to verify
assert(signature);
```